File: android-native/app/src/main/java/dev/raflos/cocina/data/remote/TokenStore.kt
package dev.raflos.cocina.data.remote

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey

class TokenStore(context: Context) {
    private val prefs: SharedPreferences by lazy {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        EncryptedSharedPreferences.create(
            context,
            "cocina_tokens",
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    }

    fun getAccess(): String? = prefs.getString(KEY_ACCESS, null)
    fun getRefresh(): String? = prefs.getString(KEY_REFRESH, null)

    fun set(access: String, refresh: String) {
        prefs.edit().putString(KEY_ACCESS, access).putString(KEY_REFRESH, refresh).apply()
    }

    fun setAccess(access: String) {
        prefs.edit().putString(KEY_ACCESS, access).apply()
    }

    fun clear() {
        prefs.edit().remove(KEY_ACCESS).remove(KEY_REFRESH).apply()
    }

    fun hasSession(): Boolean = getRefresh() != null

    companion object {
        private const val KEY_ACCESS = "access_token"
        private const val KEY_REFRESH = "refresh_token"
    }
}

File: app/src/storage/tokenStore.ts
const ACCESS_KEY = "cocina:accessToken";
const REFRESH_KEY = "cocina:refreshToken";

type Listener = () => void;
const listeners = new Set<Listener>();
const notify = () => listeners.forEach((fn) => fn());

export const tokenStore = {
  getAccess: () => window.localStorage.getItem(ACCESS_KEY),
  getRefresh: () => window.localStorage.getItem(REFRESH_KEY),
  set(access: string, refresh: string) {
    window.localStorage.setItem(ACCESS_KEY, access);
    window.localStorage.setItem(REFRESH_KEY, refresh);
    notify();
  },
  setAccess(access: string) {
    window.localStorage.setItem(ACCESS_KEY, access);
  },
  clear() {
    window.localStorage.removeItem(ACCESS_KEY);
    window.localStorage.removeItem(REFRESH_KEY);
    notify();
  },
  hasSession: () => !!window.localStorage.getItem(REFRESH_KEY),
  subscribe(fn: Listener) {
    listeners.add(fn);
    return () => {
      listeners.delete(fn);
    };
  },
};
